src/Helpers.php

```diff
@@ -156,4 +162,118 @@ public static function cleanSchemaName(string $name): string {
 		global $readableAppID;
 		return substr($name, strlen($readableAppID));
 	}
+
+	protected static function getScopeNameFromAttributeArgument(Arg $arg, string $routeName): ?string {
+		if ($arg->name->name === 'scope') {
+			if ($arg->value instanceof ClassConstFetch) {
+				if ($arg->value->class->getLast() === 'OpenAPI') {
+					return self::getScopeNameFromConst($arg->value);
+				}
+			} elseif ($arg->value instanceof String_) {
+				return $arg->value->value;
+			} else {
+				Logger::panic($routeName, 'Can not interpret value of scope provided in OpenAPI(scope: …) attribute. Please use string or OpenAPI::SCOPE_* constants');
+			}
+		}
+
+		return null;
+	}
+
+	protected static function getScopeNameFromConst(ClassConstFetch $scope): string {
+		return match ($scope->name->name) {
+			'SCOPE_DEFAULT' => 'default',
+			'SCOPE_ADMINISTRATION' => 'administration',
+			'SCOPE_FEDERATION' => 'federation',
+			'SCOPE_IGNORE' => 'ignore',
+			// Fall back for future scopes assuming we follow the pattern (cut of 'SCOPE_' and lower case)
+			default => strtolower(substr($scope->name->name, 6)),
+		};
+	}
+
+	public static function getAttributeScopes(ClassMethod|Class_|Node $node, string $annotation, string $routeName): array {
+		$scopes = [];
+
+		/** @var AttributeGroup $attrGroup */
+		foreach ($node->attrGroups as $attrGroup) {
+			foreach ($attrGroup->attrs as $attr) {
+				if ($attr->name->getLast() === $annotation) {
+					if (empty($attr->args)) {
+						$scopes[] = 'default';
+						continue;
+					}
+
+					foreach ($attr->args as $arg) {
+						$scope = self::getScopeNameFromAttributeArgument($arg, $routeName);
+						if ($scope !== null) {
+							$scopes[] = $scope;
+						}
+					}
+				}
+			}
+		}
+
+		return $scopes;
+	}
+
+	public static function getAttributeTagsByScope(ClassMethod|Class_|Node $node, string $annotation, string $routeName, string $defaultTag, string $defaultScope): array {
+		$tags = [];
+
+		/** @var AttributeGroup $attrGroup */
+		foreach ($node->attrGroups as $attrGroup) {
+			foreach ($attrGroup->attrs as $attr) {
+				if ($attr->name->getLast() === $annotation) {
+					if (empty($attr->args)) {
+						$tags[$defaultScope] = [$defaultTag];
+						continue;
+					}
+
+					$foundsTags = [];
+					$foundScopeName = null;
+					foreach ($attr->args as $arg) {
+						$foundScopeName = self::getScopeNameFromAttributeArgument($arg, $routeName);
+
+						if ($arg->name->name === 'tags') {
+							if ($arg->value instanceof Array_) {
+								foreach ($arg->value->items as $item) {
+									if ($item instanceof ArrayItem) {
+										if ($item->value instanceof String_) {
+											$foundsTags[] = $item->value->value;
+										}
+									}
+								}
+							}
+						}
+					}
+
+					if (!empty($foundsTags)) {
+						$tags[$foundScopeName ?: $defaultScope] = $foundsTags;
+					}
+				}
+			}
+		}
+
+		return $tags;
+	}
+
+	public static function collectUsedRefs(array $data): array {
```

Comment:
I fear this could run into and endless loop in case there are cyclic dependencies between schemas (which is not forbidden by json schema). Maybe checking if a ref has already been found and then skipping to check if that ref contains more refs will solve it, although that won't work easily with the recursive approach.

Reply:
It can not be infinite, we only operate on the given array and then continue with a smaller subset of the same array.

The point that could have a loop is:
```
while ($usedSchema = array_shift($usedSchemas)) {
```

But this loop is prevented by checking for existance before we add them to the `$usedSchemas`:
https://github.com/nextcloud/openapi-extractor/blob/296b99cc52a4a5bf003cd7bd7e15f196d0b284ee/generate-spec#L820-L822
```
				if (!isset($scopedSchemas[substr($newRef, strlen('#/components/schemas/'))])) {
					$usedSchemas[] = $newRef;
				}
```